### PR TITLE
feat: find_dependencies to track dependencies back to repos

### PR DESCRIPTION
**Pull request by nedbat**

### CODE REVIEW

File: edx_repo_tools/find_dependencies/find_dependencies.py [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+    repo_work.mkdir(parents=True, exist_ok=True)
+    repo_urls = set()
+    if (js_reqs := Path("package-lock.json")).exists():
+        shutil.copyfile(js_reqs, repo_work / "package-lock.json")
+        with change_dir(repo_work):
+            repo_urls.update(check_js_dependencies())
+    if (py_reqs := find_py_reqs()):
+        shutil.copyfile(py_reqs, repo_work / "base.txt")
+        with change_dir(repo_work):
+            repo_urls.update(check_py_dependencies())
```
Comment (kdmccormick) [inline edx_repo_tools/find_dependencies/find_dependencies.py:292]: Consider warning here if we suspect that the repo is a Python repo, but we can't find a suitable `py_reqs` file. Something like:
```suggestion
    if (py_reqs := find_py_reqs()):
        shutil.copyfile(py_reqs, repo_work / "base.txt")
        with change_dir(repo_work):
            repo_urls.update(check_py_dependencies())
    elif Path("setup.py").exists() or Path("setup.cfg").exists():
        print("WARNING: {repo_name} is likely a Python package, but we can't find its dependencies.")
```

**Review by kdmccormick — COMMENTED** [review]
This will be really helpful, and I found the code easy to follow. Just one request about being louder if we can't find a Python requirements list for a repo.

I would love if this printed a dependency tree to help us trace backwards when there are issues, but that sounds like a lot of work and would probably be over-engineering at this point :)

File: edx_repo_tools/find_dependencies/find_dependencies.py [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+    """Find the Python requirements file to use."""
+    for fname in PY_REQS:
+        possible_req = Path(fname)
+        if possible_req.exists():
+            return possible_req
+    return None
+
+
+def process_directory():
+    """Find all the requirements in the current directory. Returns a set of repo URLs."""
```
Comment (kdmccormick) [inline edx_repo_tools/find_dependencies/find_dependencies.py:280]: This function, critically, also copies the js/py requements list to `package-lock.json`/`base.txt`. I would note that in the docstring.

**Review by feanil — APPROVED** [review]

File: edx_repo_tools/find_dependencies/find_dependencies.py (hunk 1 above)
Comment (nedbat) [reply edx_repo_tools/find_dependencies/find_dependencies.py:302]: Thanks, good idea about the "setup.py" sniffing. I've added that in a different place in commit dea6bd9.

Printing the tree: i understand the appeal, and also the effort! One of the reasons I copied the files into the work directory was to help with analysis afterward.

**Comment by nedbat** [comment]
@kdmccormick take one more look if you would... The commit I mentioned is a little off but is corrected in another commit.  I'll squash them all before merging.

File: edx_repo_tools/find_dependencies/find_dependencies.py (hunk 1 above)
Comment (kdmccormick) [reply edx_repo_tools/find_dependencies/find_dependencies.py:303]: Looks good! And good call on copying the directory tree in.

**Review by kdmccormick — APPROVED** [review]
LGTM. Haven't thoroughly tested it but I assume you have.

**Comment by nedbat** [comment]
@kdmccormick thanks.  I have tested it, a sample run is in the README.  It takes about 90 minutes...

**Pull request merged**